### PR TITLE
Updated PowerFlex/ScaleIO gateway client with some improvements.

### PR DESCRIPTION
**Pull request by sureshanaparti**
### Description
This PR updates the PowerFlex/ScaleIO gateway client with the following improvements.

- Added connection manager to the gateway client.
- Renew the client session on '401 Unauthorized' response.
- Refactored the gateway client calls, for GET and POST methods.
- Consume the http entity content after login/(re)authentication and close the content stream if exists.
- Updated storage pool client connection timeout configuration 'storage.pool.client.timeout' to non-dynamic (It's value is picked by the ScaleIO client when it is created, which is either on the primary storage addition or management server (re)start for existing pools. In order to apply the updated config value for existing pools, it is required to restart the management server).
- Added storage pool client max connections configuration 'storage.pool.client.max.connections' (default: 100) to specify the maximum connections for the ScaleIO storage pool client.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?

- Manually tested VM / Volume life cycle operations on PowerFlex/ScaleIO storage pool.

### CODE REVIEW

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Review by rhtyd — APPROVED** [review]

File: engine/components-api/src/main/java/com/cloud/storage/StorageManager.java [hunk 1]
```diff
@@ -124,7 +124,7 @@
             "Storage",
             "60",
             "Timeout (in secs) for the storage pool client connection timeout (for managed pools). Currently only supported for PowerFlex.",
-            true,
+            false,
```
Comment (GabrielBrascher) [inline engine/components-api/src/main/java/com/cloud/storage/StorageManager.java:127]: Just out of curiosity. Why is `STORAGE_POOL_CLIENT_TIMEOUT` changed from dynamic to static?

**Review by GabrielBrascher — COMMENTED** [review]
The code overall looks good. Is this ready to review?

File: engine/components-api/src/main/java/com/cloud/storage/StorageManager.java (hunk 1 above)
Comment (sureshanaparti) [reply engine/components-api/src/main/java/com/cloud/storage/StorageManager.java:127]: > Just out of curiosity. Why is `STORAGE_POOL_CLIENT_TIMEOUT` changed from dynamic to static?

@GabrielBrascher This config is currently support for ScaleIO only. It's value is picked by the ScaleIO client when it is created, which is either on the primary storage addition or management server (re)start (for existing pools). In order to apply this config value for existing pools, it is required to restart the management server. Changing this config to 'static' prompts the operator, to restart the management server to apply the new value.

**Comment by sureshanaparti** [comment]
> The code overall looks good. Is this ready to review?

Not yet, WIP.

File: engine/components-api/src/main/java/com/cloud/storage/StorageManager.java (hunk 1 above)
Comment (GabrielBrascher) [reply engine/components-api/src/main/java/com/cloud/storage/StorageManager.java:127]: Perfect, that makes sense. Thanks for the explanation, @sureshanaparti!

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
LGTM, failures not related to this PR. @borisstoyanov are you lgtm on this since you QA'd this?
@blueorangutan package

**Comment by borisstoyanov** [comment]
Thanks @rhtyd, the testing is almost done, I'll LGMT once I'm done.

**Review by borisstoyanov — APPROVED** [review]
LGTM, manually tested storage related actions.

**Pull request merged**